File: backend/ai-service/app/schemas/session.py
from pydantic import BaseModel, Field


class SessionAttendee(BaseModel):
  user_id: str = Field(min_length=1, max_length=120)
  name: str = Field(min_length=1, max_length=120)
  email: str = Field(min_length=3, max_length=200)


class SessionAnalyzeRequest(BaseModel):
  session_id: str = Field(min_length=4, max_length=120)
  room_id: str = Field(default="")
  room_code: str = Field(default="")
  room_title: str = Field(default="", max_length=200)
  transcript_text: str = Field(default="", max_length=300000)
  sarvam_audio_url: str = Field(default="", max_length=2000)
  attendees: list[SessionAttendee] = Field(default_factory=list)


class SessionActionItem(BaseModel):
  item_id: str
  text: str
  owner_user_id: str
  owner_name: str
  priority: str = "medium"
  suggested_status: str = "todo"
  confidence: float = 0.5
  rationale: str = ""
  source_excerpt: str = ""


class SessionAnalyzeResponse(BaseModel):
  session_id: str
  summary: str
  key_points: list[str]
  action_items: list[SessionActionItem]
  transcript_source: str = "provided_text"

File: backend/node-api/src/models/Session.js
const mongoose = require("mongoose");

const sessionParticipantSchema = new mongoose.Schema(
  {
    userId: { type: String, required: true },
    name: { type: String, required: true },
    email: { type: String, required: true },
    joinedAt: { type: Date, default: Date.now },
    leftAt: { type: Date, default: null },
  },
  { _id: false },
);

const transcriptSegmentSchema = new mongoose.Schema(
  {
    speakerUserId: { type: String, default: "" },
    speakerName: { type: String, default: "" },
    text: { type: String, required: true },
    startedAtMs: { type: Number, default: null },
    endedAtMs: { type: Number, default: null },
    createdAt: { type: Date, default: Date.now },
  },
  { _id: false },
);

const sessionActionItemSchema = new mongoose.Schema(
  {
    itemId: { type: String, required: true },
    text: { type: String, required: true },
    ownerUserId: { type: String, default: "" },
    ownerName: { type: String, default: "" },
    priority: { type: String, enum: ["low", "medium", "high"], default: "medium" },
    suggestedStatus: {
      type: String,
      enum: ["todo", "in_progress", "done", "needs_clarification"],
      default: "todo",
    },
    confidence: { type: Number, default: 0.5 },
    rationale: { type: String, default: "" },
    sourceExcerpt: { type: String, default: "" },
  },
  { _id: false },
);

const sessionSchema = new mongoose.Schema(
  {
    sessionId: { type: String, required: true, unique: true, index: true },
    roomId: { type: String, default: "", index: true },
    roomCode: { type: String, default: "", index: true },
    roomTitle: { type: String, default: "" },
    createdBy: {
      userId: { type: String, required: true },
      name: { type: String, required: true },
      email: { type: String, required: true },
    },
    participants: { type: [sessionParticipantSchema], default: [] },
    status: {
      type: String,
      enum: ["live", "ended"],
      default: "live",
      index: true,
    },
    startedAt: { type: Date, default: Date.now, index: true },
    endedAt: { type: Date, default: null },
    transcriptSegments: { type: [transcriptSegmentSchema], default: [] },
    transcriptText: { type: String, default: "" },
    summary: { type: String, default: "" },
    keyPoints: { type: [String], default: [] },
    actionItems: { type: [sessionActionItemSchema], default: [] },
    processingStatus: {
      type: String,
      enum: ["pending", "analyzing", "analyzed", "failed"],
      default: "pending",
    },
    analysisError: { type: String, default: "" },
    metadata: { type: Object, default: {} },
  },
  { timestamps: true },
);

sessionSchema.index({ "participants.userId": 1, startedAt: -1 });

module.exports = mongoose.model("Session", sessionSchema);

